xtensa: esp32: Add Ethernet Support

grantramsay opened the pull request:
Hey, I had a go at adding ESP32 Ethernet support.

**Depends on this hal_espressif PR**: https://github.com/zephyrproject-rtos/hal_espressif/pull/141

Major additions:
- Added support for the [ESP32-Ethernet-Kit V1.2](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-ethernet-kit.html) board
- Added ESP32 MDIO Driver
- Added ESP32 Ethernet Driver
- Added ESP32 Ethernet Sample (possibly should be consolidated into existing networking samples?)

Ethernet and MDIO samples are working well. I've only tested on the ESP32-Ethernet-Kit.

Example output:
- Wait for successful DHCP
- Ping router
- DNS lookup google.com
- Ping google.com IP address
```
uart:~$ *** Booting Zephyr OS build zephyr-v3.1.0-4878-gda842a17fc57  ***
[00:00:00.525,000] <inf> phy_mii: PHY (1) ID 2430C54
[00:00:00.536,000] <inf> net_config: Initializing network
[00:00:00.536,000] <inf> net_config: Running dhcpv4 client...
[00:00:05.331,000] <inf> phy_mii: PHY (1) Link speed 100 Mb, full duplex
[00:00:08.351,000] <inf> net_dhcpv4: Received: 192.168.1.2
[00:00:08.351,000] <inf> net_config: IPv4 address: 192.168.1.2
[00:00:08.351,000] <inf> net_config: Lease time: 86400 seconds
[00:00:08.351,000] <inf> net_config: Subnet: 255.255.255.0
[00:00:08.351,000] <inf> net_config: Router: 192.168.1.1
uart:~$ net ping 192.168.1.1
PING 192.168.1.1
28 bytes from 192.168.1.1 to 192.168.1.2: icmp_seq=0 ttl=64 time=1 ms
28 bytes from 192.168.1.1 to 192.168.1.2: icmp_seq=1 ttl=64 time=0 ms
28 bytes from 192.168.1.1 to 192.168.1.2: icmp_seq=2 ttl=64 time=0 ms
uart:~$ net dns google.com
Query for 'google.com' sent.
dns: 172.217.167.110
dns: All results received
uart:~$ net ping 172.217.167.110
PING 172.217.167.110
28 bytes from 172.217.167.110 to 192.168.1.2: icmp_seq=0 ttl=119 time=41 ms
28 bytes from 172.217.167.110 to 192.168.1.2: icmp_seq=1 ttl=119 time=41 ms
28 bytes from 172.217.167.110 to 192.168.1.2: icmp_seq=2 ttl=119 time=41 ms
uart:~$ 
```

Not yet supported:
- RMII clock output
- Flow control (`emac_hal_enable_flow_ctrl`/`emac_ll_pause_frame_enable`)
- `CONFIG_NET_VLAN`/`ETHERNET_DUPLEX_SET`/`ETHERNET_PROMISC_MODE`
- Possibly other overlooked features

grantramsay commented on drivers/mdio/mdio_esp32.c:
```diff
@@ -0,0 +1,152 @@
+	// Poll until operation complete
+	bool success = false;
+	for (uint32_t t_us = 0; t_us < PHY_OPERATION_TIMEOUT_US; t_us += 100) {
+		k_busy_wait(100);
```
Should this loop on `k_yield()` or `k_sleep(K_USEC(...))` instead of busy wait?

grantramsay commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	struct net_pkt *pkt = net_pkt_rx_alloc_with_buffer(
+		dev_data->iface, receive_len, AF_UNSPEC, 0, K_NO_WAIT);
```
I might change this timeout from `K_NO_WAIT` to `K_MSEC(100)` to allow time for a network packet to become free.

There is also software flow control possible for RX DMA buffers that is not implemented here yet (`CONFIG_ETH_SOFT_FLOW_CONTROL`)

grantramsay commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,284 @@
+		emac_hal_iomux_init_mii();
+		emac_ll_clock_enable_mii(dev_data->hal.ext_regs);
+	} else {
+		res = EINVAL;
```
Should be `-EINVAL`

grantramsay commented on drivers/mdio/CMakeLists.txt:
```diff
@@ -4,3 +4,4 @@ zephyr_library()
 
 zephyr_library_sources_ifdef(CONFIG_MDIO_SHELL		mdio_shell.c)
 zephyr_library_sources_ifdef(CONFIG_MDIO_ATMEL_SAM	mdio_sam.c)
+zephyr_library_sources_ifdef(CONFIG_MDIO_ESP32	    mdio_esp32.c)
```
spaces -> tabs

grantramsay commented on drivers/mdio/Kconfig.esp32:
```diff
@@ -0,0 +1,9 @@
+
+config MDIO_ESP32
+	bool "ESP32 MDIO driver"
+	depends on ETH_ESP32
```
Remove invalid dependency, `MDIO_ESP32` can be used without `ETH_ESP32`

grantramsay commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
```
I'm not sure what to put in the copyright headers

grantramsay commented on boards/xtensa/esp32_ethernet_kit/doc/index.rst:
```diff
@@ -0,0 +1,6 @@
+ESP32-ETHERNET-KIT
+##############
+
+https://docs.espressif.com/projects/esp-idf/en/latest/esp32/hw-reference/esp32/get-started-ethernet-kit.html
```
Documentation for this new board is currently lacking.
Is the procedure to copy/paste/edit from a similar board?

sylvioalves replied:
Yes, you canc opy most of the setup part of the default ESP32. You can also use some specific sample code which uses ethernet as reference. If possible, add more information regarding the ethernet-devkit board itself, similar to https://docs.zephyrproject.org/3.1.0/boards/xtensa/esp_wrover_kit/doc/index.html

grantramsay commented:
I'm not sure what this twister build failure refers to:
```
FileNotFoundError: [Errno 2] No such file or directory: '_test_plan_partial.json'
```

jukkar commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,284 @@
+
+#include "eth.h"
+
+LOG_MODULE_REGISTER(eth_eth_esp32, CONFIG_ETHERNET_LOG_LEVEL);
```
I think single eth here is enough so "eth_esp32" would be fine.

jukkar commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,284 @@
+			}
+
+			if (net_recv_data(dev_data->iface, pkt) < 0) {
+				LOG_ERR("Failed to place frame in RX Queue");
```
This might print too much data if the upper layers are not yet ready to receive packets (this is not an error). I would just remove the error print or do `LOG_DBG` instead.

sylvioalves commented on boards/xtensa/esp32_ethernet_kit/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# ESP32-ETHERNET-KIT board configuration
+
+# Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
```
As far as I know, you can use you own name and email in the files you created:  
```c
/*
 * Copyright (c) 2022 Your Name <your@email>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
 ```

sylvioalves commented on boards/xtensa/esp32_ethernet_kit/esp32_ethernet_kit.dts:
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
```
Update the copyright to 2022.

sylvioalves commented on boards/xtensa/esp32_ethernet_kit/esp32_ethernet_kit_defconfig:
```diff
@@ -0,0 +1,24 @@
+CONFIG_XTENSA_USE_CORE_CRT1=n
+
+CONFIG_GPIO=y
+CONFIG_GPIO_ESP32=y
```
I think this is no longer required: `CONFIG_GPIO_ESP32=y`

sylvioalves commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,290 @@
+
+	uint32_t sent_len = emac_hal_transmit_frame(&dev_data->hal, dev_data->txb, len);
+
+	int res = len == sent_len ? 0 : -EIO;
```
I think it need one line break in here.

uLipe commented on boards/xtensa/esp32_ethernet_kit/board_init.c:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	/* Enable the Ethernet phy */
+	gpio_pin_configure(gpio, IP101GRI_RESET_N_PIN, GPIO_OUTPUT);
```
what would happen if those gpio configuration fuctions fail? 

I suggest to track when they fail, otherwise it would be hard to debug if external phy fails because of something related to reset.

uLipe commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,290 @@
+
+	k_sem_init(&dev_data->int_sem, 1, 1);
+
+	periph_module_enable(PERIPH_EMAC_MODULE);
```
I think you should enable ethernet clock using esp32 clock control API.

grantramsay replied:
Can you check that I've done this correctly

uLipe commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,290 @@
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+	int res;
+
+	k_sem_init(&dev_data->int_sem, 1, 1);
```
Is this a binary semaphore? If so it should be initialized with 0, right?

uLipe commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,290 @@
+	bool reset_success = false;
+
+	for (uint32_t t_ms = 0; t_ms < MAC_RESET_TIMEOUT_MS; t_ms += 10) {
+		k_sleep(K_MSEC(10));
```
I'm wondering here, since the eth init runs at system initialization context if there is some risk of k_sleep fail here, it would not be better to use busy_wait instead?

uLipe commented on drivers/ethernet/eth_esp32.c:
```diff
@@ -0,0 +1,290 @@
+	}
+	emac_hal_set_address(&dev_data->hal, dev_data->mac_addr);
+
+	k_tid_t tid = k_thread_create(
```
what happens if thread creation fails? Suggest to check if creation was succesful.

grantramsay replied:
zephyr doesn't offer a way to check thread creation success.
However I've modified this thread to have the `K_ESSENTIAL` attribute

uLipe commented on drivers/mdio/mdio_esp32.c:
```diff
@@ -0,0 +1,154 @@
+	struct mdio_esp32_dev_data *const dev_data = dev->data;
+	int res;
+
+	k_sem_init(&dev_data->sem, 1, 1);
```
same here, is this a binary semaphore? if so it should be initialized to 1 key and 0 keys available

grantramsay replied:
The other semaphore (eth_esp32.c) was incorrectly initialized. However I believe this one is correct.
In this case there is one device/resource, and it is ready to be used/taken after initialization

uLipe replied:
Yes now I see it, just a minor question (feel free to resolve), why not use a mutex instead? Do you believe the priority inversion risk here is negligible?

grantramsay replied:
It could be a mutex, TBH I used this because it was similar to how the SAM MDIO driver did it (mdio_sam.c).
PHY operations are pretty fast (<1ms) so shouldn't block for very long, and most are likely done from a single thread (except MDIO_SHELL which seems more for debug)

uLipe commented on drivers/mdio/mdio_esp32.c:
```diff
@@ -0,0 +1,154 @@
+		goto err;
+	}
+
+	periph_module_enable(PERIPH_EMAC_MODULE);
```
it should be enabled using clock control API.

uLipe reviewed (changes requested):
Great work!

Left some minor requests / questions.

grantramsay commented:
Just realized I've missed setting the net_if link up/down when Ethernet is connected/disconnected.
Adding a small commit to add this functionality...

jukkar commented on boards/xtensa/esp32_ethernet_kit/board_init.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
+ * Copyright (c) 2022 Grant Ramsay <grant.ramsay@hotmail.com>
```
Doing this here looks a bit weird, I mean that if your original code is copied from Expressif, then you should not remove their copyright. If you by accident had the wrong copyright, please fix the sources in previous commits.

jukkar commented on boards/xtensa/esp32_ethernet_kit/esp32_ethernet_kit_defconfig:
```diff
@@ -16,7 +16,6 @@ CONFIG_UART_CONSOLE=y
 CONFIG_XTENSA_USE_CORE_CRT1=n
```
There are multiple tidy commits, perhaps merge them together?

grantramsay committed:
boards: xtensa: Add esp32_ethernet_kit board

Adds support for the esp32_ethernet_kit board.
esp32_ethernet_kit board is useful for testing ESP32 Ethernet

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

grantramsay committed:
boards: xtensa/esp32_ethernet_kit: Init Ethernet phy

Initialises the Ethernet phy hardware.
Sets phy RESET_N pin high shortly after boot

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

grantramsay committed:
drivers: mdio: Add ESP32 MDIO driver

The MDIO driver is required to for ESP32 Ethernet

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

grantramsay committed:
drivers: ethernet: Add ESP32 Ethernet driver

This is to enable ESP32 Ethernet support

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

grantramsay committed:
boards: xtensa/esp32_ethernet_kit: Add Ethernet support

Configures ESP32 Ethernet support for the esp32_ethernet_kit board

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

grantramsay committed:
samples: boards/esp32: Add ESP32 Ethernet sample

A sample project for ESP32 Ethernet testing.
Possibly this should be consolidated into existing networking samples

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

sylvioalves commented on boards/xtensa/esp32_ethernet_kit/doc/index.rst:
```diff
@@ -0,0 +1,544 @@
+Enabling Ethernet
+-----------------
+
+Enable Ethernet MAC, PHY and MDIO; add these to your device tree overlay::
```
Nitpick: duplicated colon

grantramsay replied:
I changed this but will have to change it back...
`::` is used for code blocks, it renders as a single `:`
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

sylvioalves commented on boards/xtensa/esp32_ethernet_kit/doc/index.rst:
```diff
@@ -0,0 +1,544 @@
+        status = "okay";
+    };
+
+Enable Ethernet in KConfig::
```
Nitpick: duplicated colon

sylvioalves commented:
@grantramsay Just some comments, LGTM.

grantramsay commented on boards/xtensa/esp32_ethernet_kit/board_init.c:
```diff
@@ -0,0 +1,28 @@
+
+#include <zephyr/drivers/gpio.h>
+
+#define IP101GRI_RESET_N_PIN	5
```
I _think_ this is correct. The phy_mii driver does not have dts bindings for a reset pin, as it's not part of the standard.
AFAIK this is a workaround for the wrover module using GPIO0 as a bootstrapping pin:

> To prevent the power-on state of the GPIO0 from being affected by the clock output on the PHY side, the RESET_N signal to PHY defaults to low, turning the clock output off. After power-on you can control RESET_N with GPIO5 to turn the clock output on

grantramsay committed:
west.yml: update hal_espressif to add ESP32 Ethernet support

ESP32 Ethernet support required hal_espressif modifications:
Espressif Ethernet MAC HAL and pinctrl definitions

Signed-off-by: Grant Ramsay <grant.ramsay@hotmail.com>

fabiobaltieri commented:
Hey could you rework the board images, reduce the resolution a bit and convert the photo ones to jpg? Trying to keep the repo size under control

grantramsay commented:
> Hey could you rework the board images, reduce the resolution a bit and convert the photo ones to jpg? Trying to keep the repo size under control

@fabiobaltieri I carried out these steps and forced pushed the old images out of the git history:

``` shell
$ cd boards/xtensa/esp32_ethernet_kit/doc/img
$ du -d0 -h
1.1M   .
$ mogrify -resize 50% *.jpg
$ # Set transparent background to white
$ mogrify -background white -alpha remove -alpha off -resize 50% -format jpg *.png
$ rm *.png
$ du -d0 -h
236K    .
```

grantramsay commented:
@sylvioalves could you give this another review?
Just reduced image sizes and corresponding readme updates since last review

grantramsay commented:
Cool, I think this is ready to merge. I'm not authorized to do so. Can someone with permissions merge it?
